File: Scala/src/main/scala/AlgorithmicHeights/PS.scala
package AlgorithmicHeights

object PS {
  import scala.annotation.tailrec

  private def convertToIntArray(line: String): Array[Int] = line.split(" ").map(_.toInt)

  private def getIndexOfParentChildrenMinimum(array: Array[Int], parentIx: Int, size: Int): Int = {
    val leftChildIx: Int = 2 * parentIx + 1
    val rightChildIx: Int = leftChildIx + 1
    val minIndex: Int = if (leftChildIx < size && array(leftChildIx) < array(parentIx)) leftChildIx else parentIx
    if (rightChildIx < size && array(rightChildIx) < array(minIndex)) rightChildIx else minIndex
  }

  private def swapElems(array: Array[Int], ix: Int, jy: Int): Unit = {
    val elem: Int = array(ix)
    array(ix) = array(jy)
    array(jy) = elem
  }

  private def siftDown(array: Array[Int], parentIx: Int, n: Int): Unit = {
    @tailrec
    def loop(currentParentIx: Int): Unit = {
      val minIx: Int = getIndexOfParentChildrenMinimum(array, currentParentIx, n)
      if (minIx == currentParentIx) ()
      else {
        swapElems(array, minIx, currentParentIx)
        loop(minIx)
      }
    }

    loop(parentIx)
  }

  private def heapify(array: Array[Int], n: Int): Unit =
    (n / 2 - 1 to 0 by -1).foreach(parentIx => siftDown(array, parentIx, n))

  def partialSort(array: Array[Int], size: Int, k: Int): Vector[Int] = {
    heapify(array, size)
    (size until (size - k) by -1).foreach{
      n =>
        swapElems(array, 0, n - 1)
        siftDown(array, 0, n - 1)
    }
    array.takeRight(k).reverseIterator.toVector
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val array: Array[Int] = convertToIntArray(reader.next())
    val k: Int = reader.next().toInt
    val result: Vector[Int] = partialSort(array, n, k)
    println(result.mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/ps.ml
let convert_to_ints (line : string) : int array =
    Array.of_list @@ List.map int_of_string Str.(line |> split (regexp " "))


let get_index_of_parent_children_minimum (arr : int array) (parent_ix : int) (size : int) : int =
    let left_child_ix = 2 * parent_ix + 1 in
    let right_child_ix = left_child_ix + 1 in
    let min_ix = if left_child_ix < size && arr.(left_child_ix) < arr.(parent_ix) then left_child_ix else parent_ix in
    if right_child_ix < size && arr.(right_child_ix) < arr.(min_ix) then right_child_ix
    else min_ix


let swap arr i j =
    let elem = arr.(i) in
    arr.(i) <- arr.(j);
    arr.(j) <- elem


let sift_down (arr : int array) (parent_ix : int) (n : int) : unit =
    let rec loop current_parent_ix =
        let min_ix = get_index_of_parent_children_minimum arr current_parent_ix n in
        if min_ix <> current_parent_ix then
            (swap arr min_ix current_parent_ix;
            loop min_ix) in
    loop parent_ix


let heapify (arr : int array) (n : int) : unit =
    for parent_ix = (n / 2 - 1) downto 0 do
        sift_down arr parent_ix n;
    done


let partialsort (arr : int array) (size : int) (k : int) : int list =
    heapify arr size;
    for n = size downto (size - k + 1) do
        swap arr 0 (n - 1);
        sift_down arr 0 (n - 1)
    done;
    List.rev @@ Array.to_list @@ Array.sub arr (size - k) k


let () =
    let n = read_int () in
    let arr = convert_to_ints (read_line ()) in
    let k = read_int () in
    let result = partialsort arr n k in
    result |> List.map string_of_int |> String.concat " " |> print_endline
